subsys/settings/src/settings_zms.c

```diff
@@ -237,9 +294,22 @@ static int settings_zms_save(struct settings_store *cs, const char *name, const
 	/* MSB is always 1 */
 	name_hash |= BIT(31);
 
-	/* Let's find out if there are hash collisions in the storage */
+#ifdef CONFIG_SETTINGS_ZMS_NAME_CACHE
+	uint8_t cache_flags = 0;
+
+	cache_flags = settings_zms_cache_match(cf, name_hash & ZMS_HASH_MASK);
+	if (ZMS_CACHE_EXIST(cache_flags) && !ZMS_CACHE_HAS_COLLISION(cache_flags)) {
```

Comment:
I think this code is wrong. Suppose that you have 2 keys that have the same hash: A (collision number 0) and B (collision number 1). If you call `save_one(A)`, you will add an entry to the name cache with flag HAS_COLLISION cleared. If you then call `save_one(B)`, this `if` will be satisfied and you will write the value for the wrong key.